### ISSUE TITLE
Performance insanity!

### ISSUE THREAD

**Issue by linas** [opencog/atomspace-rocks#11]
On small datasets and on synthetic benchmarks, the RocksDB backend is 2x or 3x faster than the Postgres backend. This is not surprising, given the complexity of mapping the AtomSpace to SQL, together with the overhead of client-server communications. However, in a real-world use-case, RocksDB is 9x slower than Postgres. Why? What is going wrong?  Is this related to issues #10 and #9? Will fixes to those also fix this?

From the learning project. RocksDB:
```
(define wsv (make-shape-vec-api))
(define wss (add-pair-stars wsv))
(define wst (batch-transpose wss))
(wst 'mmt-marginals)
...
Stored 40000 of 510993 left-wilds in 101 secs (396 pairs/sec)
Stored 80000 of 510993 left-wilds in 356 secs (112 pairs/sec)
Stored 120000 of 510993 left-wilds in 607 secs (66 pairs/sec)
Stored 160000 of 510993 left-wilds in 865 secs (46 pairs/sec)
Stored 200000 of 510993 left-wilds in 1075 secs (37 pairs/sec)
Stored 240000 of 510993 left-wilds in 1352 secs (30 pairs/sec)
Stored 280000 of 510993 left-wilds in 1591 secs (25 pairs/sec)
Stored 320000 of 510993 left-wilds in 1841 secs (22 pairs/sec)
Stored 360000 of 510993 left-wilds in 2076 secs (19 pairs/sec)
Stored 400000 of 510993 left-wilds in 2296 secs (17 pairs/sec)
Stored 440000 of 510993 left-wilds in 2584 secs (15 pairs/sec)
Stored 480000 of 510993 left-wilds in 2871 secs (14 pairs/sec)
Done storing 510993 left-wilds in 19969 secs
```
Freakin disaster. That works out to 5.5 hours, or 26 pairs/second. Compare this to exactly the same dataset on Postgres:
```
Stored 40000 of 510993 left-wilds in 170 secs (235 pairs/sec)
Stored 80000 of 510993 left-wilds in 168 secs (238 pairs/sec)
Stored 120000 of 510993 left-wilds in 188 secs (213 pairs/sec)
Stored 160000 of 510993 left-wilds in 177 secs (226 pairs/sec)
Stored 200000 of 510993 left-wilds in 171 secs (234 pairs/sec)
Stored 240000 of 510993 left-wilds in 165 secs (242 pairs/sec)
Stored 280000 of 510993 left-wilds in 164 secs (244 pairs/sec)
Stored 320000 of 510993 left-wilds in 176 secs (227 pairs/sec)
Stored 360000 of 510993 left-wilds in 163 secs (245 pairs/sec)
Stored 400000 of 510993 left-wilds in 189 secs (212 pairs/sec)
Stored 440000 of 510993 left-wilds in 187 secs (214 pairs/sec)
Stored 480000 of 510993 left-wilds in 185 secs (216 pairs/sec)
Done storing 510993 left-wilds in 2246 secs
```
Postgres took 37 minutes to store the same dataset. What's the problem, here?  How can we fix this?

**Comment by linas** [comment]
FYI, re-opening the same dataset, and storing exactly the same data back again, gives an insanely different result:
```
...
Stored 440000 of 510993 left-wilds in 3 secs (13333 pairs/sec)
Stored 480000 of 510993 left-wilds in 4 secs (10000 pairs/sec)
Done storing 510993 left-wilds in 43 secs
```
Wow!? This suggests that the first time, the Rocks bottleneck was that, perhaps, the indexes were being recomputed with each store!?

**Comment by linas** [comment]
Performance insight: if rocks is asked to write the same records over and over, e.g. the non-interleaved `MultiDeleteUTest`, then it runs 30x faster than Postgres. However, when deletes and writes of the same records are interleaved, all performance gains are lost, and postgres and rocks run about the same speed. (rocks is 30% faster, but comparable)

**Comment by linas** [comment]
This behavior appears to have been fixed in eedfbc78c4a056fa852707eef0f1220a3ee2c6fe ... although I don't see how.

That commit redesigned the incoming set layout....

**Comment by linas** [comment]
Closing. Yesterday's run shows:
```         
Stored 80000 of 485786 left-wilds in 14 secs (2857 pairs/sec)        
Stored 120000 of 485786 left-wilds in 16 secs (2500 pairs/sec)     
Stored 160000 of 485786 left-wilds in 16 secs (2500 pairs/sec)    
Stored 200000 of 485786 left-wilds in 14 secs (2857 pairs/sec)    
Stored 240000 of 485786 left-wilds in 16 secs (2500 pairs/sec)    
Stored 280000 of 485786 left-wilds in 17 secs (2353 pairs/sec)  
Stored 320000 of 485786 left-wilds in 18 secs (2222 pairs/sec)   
Stored 360000 of 485786 left-wilds in 17 secs (2353 pairs/sec)  
Stored 400000 of 485786 left-wilds in 14 secs (2857 pairs/sec)                  
Stored 440000 of 485786 left-wilds in 15 secs (2667 pairs/sec)                  
Stored 480000 of 485786 left-wilds in 15 secs (2667 pairs/sec)
Done storing 485786 left-wilds in 189 secs
```

This is a vast improvement; it is now 2246 / 189 = 12x faster than postgres! Yay!!
